### ISSUE TITLE
Creating a 'plugin' for Kluster

### ISSUE THREAD

**Issue by davesteps** [noaa-ocs-hydrography/kluster#83]
Hi @ericgyounkin,
As you're aware we are developing algorithms/models for cleaning soundings from multibeam data. I noticed #23 and we have been mainly using GSF as a format to share data so being able to read GSF files would great. We are always exploring tools that might help us iterate our algorithms faster. The inputs to the models are essentially dask dataframes of x,y,z data. The two things I would be looking to try first would be:

- Apply an algorithm to a full survey (all loaded lines)
- Apply an algorithm to selected subset (same subset that is in the 3d points view)

My feeling is that it shouldn't be too hard as you are using xarray datasets which are easily converted to dask data frames. Any pointers you can give to help us try this out, such us code structure/architecture, how to interact with the data or create some basic gui elements, would be much appreciated. One specific question I have:

- Is there a way to use the GUI in debug or interactive mode IE so we can see in a console what is being run?

Thanks

**Comment by ericgyounkin** [comment]
So I think there are a few things here.  First, I would like to export/read GSF.  I don't know much about the format, and I'm not sure where to start.  I see the UKHO gsfpy, but it doesn't work on Windows?  And there is the schwehr implementation I see referenced in the readme.  Any advice for where to start learning?

Getting reading GSF to work would depend on what is in the format, specifically the metadata.  I use things like the sonar serial number to catalogue files, we'd just have to figure out what is generally in a GSF file.  Starting with reading GSF that Kluster generates would probably make all that easier.

One item that is currently on my list is to implement a filter module/UI for doing basically what you suggest here, I think.  You can see that item [here](https://trello.com/c/gk9p8Qhd/30-filtering).  It is basically what you describe above, run algorithms from simple (remove all beams greater than angle x from Line/Dataset/Points Subset) to plugin-able, where you drop your custom template .py file in a directory, and it shows up as an option in the dropdown in the tool.  

I do have [code examples](https://github.com/noaa-ocs-hydrography/kluster/tree/master/examples) in the repo, but maybe not one that matches your need.  I can look at building one.

There is not currently a debug mode, but I don't think that would be too difficult to implement.  Assuming we are just talking about spitting out some text showing the methods and arguments corresponding to what is shown on screen.

**Comment by ericgyounkin** [comment]
[More examples](https://github.com/noaa-ocs-hydrography/kluster/tree/kluster_0_8_10/examples) in the new branch to show the use of the polygon subset and some data structures exploration.

**Comment by ericgyounkin** [comment]
I should also mention that all the API docs should be up to date, if you want to read about what functions/methods are available.  

You could start with the FQPR object, which is what you get on reload_data, and is the parent object for all the data.

https://kluster.readthedocs.io/en/latest/api/HSTB.kluster.fqpr_generation.Fqpr.html#HSTB.kluster.fqpr_generation.Fqpr

**Comment by davesteps** [comment]
Ok thanks for the info @ericgyounkin. I have managed to get it running in debug mode using Pycharm which is great to figure out how things are working (There was a slight issue with finding the qgis path in `_qt.py` as I'm in Linux). 

The idea of a template .py so you can plugin you own filtering algos sounds like a great idea would happily provide input and test this. 

Regarding the GSF issue appreciate it is not a simple thing. In the first instance will probably try to implement writing out to a GSF using the UKHO/gsfpy library as this will be very useful for us, happy to open a PR if we mange it, appreciate it will only be a linux solution at this stage.

**Comment by ericgyounkin** [comment]
Sounds good @davesteps.  

I am curious about your _qt.py fix.  I did not have any issues with qgis paths in my Ubuntu test environment.  I am curious what you had to change to make it work.

I went into `main()` in `kluster_main.py` and added the following, just to show what my linux paths look like.

```
print(found_path)
/home/eyou102/miniconda3/envs/kluster_test/share/qgis/python/qgis
print(plugin_dir)
/home/eyou102/miniconda3/envs/kluster_test/share/qgis/python/plugins
print(prefix_dir)
/home/eyou102/miniconda3/envs/kluster_test/share/qgis/python/qgis/qgis
```

My prefix_dir doesn't exist, which is odd, but the other paths do.  And I am able to change the background layers to satellite, etc., so I think the plugins work.

Do we have different paths?

**Comment by davesteps** [comment]
Yes it is a bit strange it is only when I run from Pycharm. It works fine when I run using `python -m HSTB.kluster` or `python HSTB/kluster/gui/kluster_main.py` however when I run `kluster_main.py` from within my pycharm it returns:

```
Traceback (most recent call last):
  File "/home/david/kluster/HSTB/kluster/gui/kluster_main.py", line 25, in <module>
    from HSTB.kluster.gui import dialog_vesselview, kluster_explorer, kluster_project_tree, kluster_3dview_v2, \
  File "/home/david/kluster/HSTB/kluster/gui/kluster_2dview.py", line 6, in <module>
    raise EnvironmentError('QGIS not found, Unable to run Kluster')
OSError: QGIS not found, Unable to run Kluster
```

I have just found a workaround for now by adding `sys.path.append("/home/david/miniconda3/envs/kluster_test/share/qgis/python")` to the top of `_qt.py`. This allows me to run in debug mode in pycharm.

**Comment by ericgyounkin** [comment]
@davesteps I thought I'd share something I'm working on for Kluster 0.9.  I've developed a plugin architecture for filtering, that you can run on a dataset, line, or in the Points View.  The filter module will find all filter .py files on the file system, so you can drop in a new .py file and run the filter in the GUI.  You can use the 'controls' attribute of the Filter to build custom GUI controls, if you need the user to specify 'minimum angle' or something to the filter as well.

I was wondering if you might have time to briefly look at one of my examples in the branch.

https://github.com/noaa-ocs-hydrography/kluster/blob/kluster_0_9_0/HSTB/kluster/plugins/filters/filter_by_angle.py

Which inherits from the base class here.

https://github.com/noaa-ocs-hydrography/kluster/blob/kluster_0_9_0/HSTB/kluster/modules/filter.py#L201

It illustrates the filter design.  You would write a simple function as shown in filter_by_angle to get a new sounding status for each sounding, and save that new status.  

Do you think that makes sense?  I just want to get a gut reaction before I keep going.  Thanks.

**Comment by davesteps** [comment]
Thanks @ericgyounkin, This sounds very useful, I'll find some time to have a look and try this out. It comes at a good time as we have been experimenting with incorporating our filtering algorithms to run on the 3d points view.

**Comment by ericgyounkin** [comment]
@davesteps sounds good, I should release 0.9 here in the next week or so.

**Comment by davesteps** [comment]
Hi @ericgyounkin, My colleague (@asmith26) and I have been experimenting with the filter plugins and it looks like a good solution so many thanks! The whole survey/single line/ selected points option is really useful, and specifically something we were after. The Filter class provides a straight forward solution so is fairly easy to adapt. It might be a nice to add a notebook with an example to get folks started on creating a new filter, included might be how we go about using the existing dask cluster client that kluster has created. Perhaps this is something we can do.

We were wondering if you be open to the idea of having a call with us at some point? We would be interested general chat about Kluster what are your roadmap for it? Are there avenues for collaboration etc? We could share what we have done so far re the filter integration.
 
Thanks Again!

**Comment by ericgyounkin** [comment]
To resolve this thread, I released Kluster 0.9.0 which includes support for running customized filters.  See https://kluster.readthedocs.io/en/latest/cleaning/index.html#filter
